app/models/grda_warehouse/auth_policies/policy_provider.rb

```diff
@@ -0,0 +1,86 @@
+require 'memery'
+
+class GrdaWarehouse::AuthPolicies::PolicyProvider
+  include Memery
+  attr_reader :user
+
+  def initialize(user)
+    @user = user
+  end
+
+  # TODO: START_ACL remove after ACL migration is complete
+  ###
+  # The legacy permission policy only checks the permissions on a user's roles and assumes the loaded record is visible
+  # to the user already. This is a dangerous assumption so it's required to confirm the record is authorized by
+  # setting legacy_implicitly_assume_authorized_access to true.
+  # Note this is only relevant for users where using_acls is false
+  ###
+  attr_accessor :legacy_implicitly_assume_authorized_access
+
+  def for_client(client_or_id)
+    client_id = id_from_arg(client_or_id, GrdaWarehouse::Hud::Client)
+    if user.using_acls?
+      client_project_policy(client_id)
+    else
+      # TODO: START_ACL remove after ACL migration is complete
+      handle_legacy_unauthorized unless legacy_implicitly_assume_authorized_access
+      legacy_user_role_policy
+    end
+  end
+
+  def for_patient(patient)
+    for_client(patient.client_id)
+  end
+
+  memoize def for_project(project_or_id)
+    project_id = id_from_arg(project_or_id, GrdaWarehouse::Hud::Project)
+    if user.using_acls?
+      GrdaWarehouse::AuthPolicies::ProjectPolicy.new(user: user, project_id: project_id)
+    else
+      # TODO: START_ACL remove after ACL migration is complete
+      handle_legacy_unauthorized unless legacy_implicitly_assume_authorized_access
+      legacy_user_role_policy
+    end
+  end
+
+  protected
+
+  def handle_legacy_unauthorized
+    raise "legacy authorization not performed"
+  end
+
+  memoize def client_project_policy(client_id)
+    project_policies = visible_client_project_ids(client_id).map do |project_id|
+      for_project(project_id)
+    end
+    GrdaWarehouse::AuthPolicies::AnyPolicy.new(project_policies)
+  end
+
+  # TODO: START_ACL remove after ACL migration is complete
+  memoize def legacy_user_role_policy
+    GrdaWarehouse::AuthPolicies::LegacyUserRolePolicy.new(user: user)
+  end
+
+  # Needs review
+  def visible_client_project_ids(client_id)
+    p_t = GrdaWarehouse::Hud::Project.arel_table
+    GrdaWarehouse::Hud::Enrollment
+      .visible_to(user, client_ids: [client_id])
+      .joins(:project).order(p_t[:id]).pluck(p_t[:id])
```

Comment:
@eanders this probably isn't right. Can you point me in the right direction? Given an acl-using user, how should I find the set of authorized projects relevant to the client_id?

Reply:
@ttoomey I think we'll need to crib from `drivers/client_access_control/extensions/grda_warehouse/hud/client_extension.rb` and `drivers/client_access_control/app/models/client_access_control/enrollment_arbiter.rb`.  Specifically, I think `ClientAccessControl::EnrollmentArbiter#clients_destination_visible_to` and `ClientAccessControl::EnrollmentArbiter#clients_source_visible_to`.  The trick is that we've never done per-client per-field permissions, so there's some ambiguity (or not yet determined functionality) around how we handle things like releases of information.  If a client has an ROI on file, should you be able to see their name even if otherwise you wouldn't be able to even see their record at all?

Reply:
I _think_ this ends up being something like:
```
arbiter = GrdaWarehouse::Config.arbiter_class.new(user)
arbiter..enrollments_visible_to(user, client_ids: some_set_of_client_ids_for_performance).
  joins(:project).
  pluck(p_t[:id])
```

The issue with the above is that it's missing the ability to pass in the related permission.  Maybe that's ok because you're checking the permission against the `project_id` in `client_project_policy` but it negates the ROI since the ROI doesn't automatically add projects to your list.  I think we need a method where you can ask "For which clients do I have this permission?"

I'll push a possible solution to the enrollment arbiter class.